### PR TITLE
feat: add hooks to product module's workflows

### PR DESCRIPTION
**Pull request by thetutlage**
Fixes: FRMW-2641

In this PR we have introduce hooks within the product module's workflows. Following is the collection of workflows and hooks they expose.

## Create action workflows

<dl>

<dt> createCollectionsWorkflow </dt>
<dd>

- Hook: `collectionsCreated`
- Input: `{ collections: ProductCollectionDTO[], additional_data?: AdditionalData }`

</dd>

<dt> createProductOptionsWorkflow </dt>
<dd>

- Hook: `productOptionsCreated`
- Input: `{ product_options: ProductOptionDTO[], additional_data?: AdditionalData }`

</dd>

<dt> createProductTagsWorkflow </dt>
<dd>

- Hook: `productTagsCreated`
- Input: `{ product_tags: ProductTagDTO[], additional_data?: AdditionalData }`

</dd>

<dt> createProductTypesWorkflow </dt>
<dd>

- Hook: `productTypesCreated`
- Input: `{ product_types: ProductTypeDTO[], additional_data?: AdditionalData }`

</dd>

<dt> createProductVariantsWorkflow </dt>
<dd>

- Hook: `productVariantsCreated`
- Input: `{ product_variants: ProductVariantDTO[], additional_data?: AdditionalData }`

</dd>

<dt> createProductsWorkflow </dt>
<dd>

- Hook: `productsCreated`
- Input: `{ products: ProductDTO[], additional_data?: AdditionalData }`

</dd>

</dl>

## Update action workflows
<dl>

<dt> updateCollectionsWorkflow </dt>
<dd>

- Hook: `collectionsUpdated`
- Input: `{ collections: ProductCollectionDTO[], additional_data?: AdditionalData }`

</dd>

<dt> updateProductOptionsWorkflow </dt>
<dd>

- Hook: `productOptionsUpdated`
- Input: `{ product_options: ProductOptionDTO[], additional_data?: AdditionalData }`

</dd>

<dt> updateProductTagsWorkflow </dt>
<dd>

- Hook: `productTagsUpdated`
- Input: `{ product_tags: ProductTagDTO[], additional_data?: AdditionalData }`

</dd>

<dt> updateProductTypesWorkflow </dt>
<dd>

- Hook: `productTypesUpdated`
- Input: `{ product_types: ProductTypeDTO[], additional_data?: AdditionalData }`

</dd>

<dt> updateProductVariantsWorkflow </dt>
<dd>

- Hook: `productVariantsUpdated`
- Input: `{ product_variants: ProductVariantDTO[], additional_data?: AdditionalData }`

</dd>

<dt> updateProductsWorkflow </dt>
<dd>

- Hook: `productsUpdated`
- Input: `{ products: ProductDTO[], additional_data?: AdditionalData }`

</dd>

</dl>

## Delete action workflows
<dl>

<dt> deleteCollectionsWorkflow </dt>
<dd>

- Hook: `collectionsDeleted`
- Input: `{ ids: string[] }`

</dd>

<dt> deleteProductOptionsWorkflow </dt>
<dd>

- Hook: `productOptionsDeleted`
- Input: `{ ids: string[] }`

</dd>

<dt> deleteProductTagsWorkflow </dt>
<dd>

- Hook: `productTagsDeleted`
- Input: `{ ids: string[] }`

</dd>

<dt> deleteProductTypesWorkflow </dt>
<dd>

- Hook: `productTypesDeleted`
- Input: `{ ids: string[] }`

</dd>

<dt> deleteProductVariantsWorkflow </dt>
<dd>

- Hook: `productVariantsDeleted`
- Input: `{ ids: string[] }`

</dd>

<dt> deleteProductsWorkflow </dt>
<dd>

- Hook: `productsDeleted`
- Input: `{ ids: string[] }`

</dd>

</dl>

### CODE REVIEW

**Comment by thetutlage** [comment]
Heads up @olivermrbl @adrien2p. This is approach I am planning to take.

**Comment by olivermrbl** [comment]
LGTM, however, I would argue we should start with Products specifically. Collections, types, tags, categories are less important compared to the two former.

But perhaps, this was more for demonstration purposes?

**Comment by thetutlage** [comment]
> LGTM, however, I would argue we should start with Products specifically. Collections, types, tags, categories are less important compared to the two former.

I thought by product we mean, every workflow related to products inside the `core-flows/src/products/workflows` directory. But anyways, I have added hooks to all these workflows.

> But perhaps, this was more for demonstration purposes?
Yes, it was

File: packages/core/core-flows/src/product/workflows/create-collections.ts [hunk 1]
```diff
@@ -2,18 +2,28 @@ import { ProductTypes } from "@medusajs/types"
...
 export const createCollectionsWorkflowId = "create-collections"
 export const createCollectionsWorkflow = createWorkflow(
   createCollectionsWorkflowId,
-  (
-    input: WorkflowData<WorkflowInput>
-  ): WorkflowResponse<ProductTypes.ProductCollectionDTO[]> => {
-    return new WorkflowResponse(createCollectionsStep(input.collections))
+  (input: WorkflowData<WorkflowInput>) => {
+    const collections = createCollectionsStep(input.collections)
+    const collectionsStepCreated = createHook("collectionsStepCreated", {
```
Comment (adrien2p) [inline packages/core/core-flows/src/product/workflows/create-collections.ts]: suggestion: in some cases the hook name include `step` in other no, should we unify?

Comment (olivermrbl) [inline packages/core/core-flows/src/product/workflows/create-collections.ts]:
```suggestion
    const collectionsCreated = createHook("collectionsCreated", {
```

File: packages/core/core-flows/src/product/workflows/update-products.ts [hunk 2]
```diff
@@ -248,9 +247,14 @@ export const updateProductsWorkflow = createWorkflow(
     })
 
     dismissRemoteLinkStep(toDeleteSalesChannelLinks)
-
     createRemoteLinkStep(salesChannelLinks)
 
-    return new WorkflowResponse(updatedProducts)
+    const productsUpdated = createHook("productsUpdated", {
+      products: updatedProducts,
```
Comment (olivermrbl) [inline packages/core/core-flows/src/product/workflows/update-products.ts:255]: question: Should this have additional data too?

For example, you might want to update the associated `brand_id` of the product.

File: packages/core/core-flows/src/product/workflows/update-collections.ts [hunk 3]
```diff
@@ -2,23 +2,33 @@ import { ProductTypes } from "@medusajs/types"
...
   updateCollectionsWorkflowId,
-  (
-    input: WorkflowData<WorkflowInput>
-  ): WorkflowResponse<ProductTypes.ProductCollectionDTO[]> => {
-    return new WorkflowResponse(updateCollectionsStep(input))
+  (input: WorkflowData<WorkflowInput>) => {
+    const updatedCollections = updateCollectionsStep(input)
+    const collectionsUpdated = createHook("collectionsUpdated", {
+      selector: input.selector,
+      update: input.update,
```
Comment (adrien2p) [inline packages/core/core-flows/src/product/workflows/update-collections.ts]: Q: do we have to provide the input `selector|update` here?  what was the thinking? Maybe the update in order to try to find what has changed after update right? but the selector I am not sure it is needed, it is more related to what the user want to receive as the result

Comment (adrien2p) [reply packages/core/core-flows/src/product/workflows/update-collections.ts]: Same question for the other workflows

**Review by adrien2p — DISMISSED** [review]

File: packages/core/core-flows/src/product/workflows/update-collections.ts (hunk 3 above)
Comment (olivermrbl) [inline packages/core/core-flows/src/product/workflows/update-collections.ts]: thought: Don't think we need these. What's your thinking?

File: packages/core/core-flows/src/product/workflows/create-collections.ts (hunk 1 above)
Comment (srindom) [reply packages/core/core-flows/src/product/workflows/create-collections.ts]: would suggest that we don't use the notion of a step in the hook names

Comment (olivermrbl) [reply packages/core/core-flows/src/product/workflows/create-collections.ts]: > would suggest that we don't use the notion of a step in the hook names

+1

File: packages/core/core-flows/src/product/workflows/update-collections.ts (hunk 3 above)
Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/update-collections.ts]: Yeah. I have no suggestions on this. If `selector` `update` seems redundant then we can remove them. CC @olivermrbl what you think?

File: packages/core/core-flows/src/product/workflows/update-products.ts (hunk 2 above)
Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/update-products.ts:255]: Yes. It should. Maybe I missed it

File: packages/core/core-flows/src/product/workflows/update-collections.ts (hunk 3 above)
Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/update-collections.ts]: Yeah @adrien2p pointed the same. Lemme remove them

File: packages/core/core-flows/src/product/workflows/create-collections.ts (hunk 1 above)
Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/create-collections.ts]: > suggestion: in some cases the hook name include step in other no, should we unify?

Was a typo :)

Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/create-collections.ts]: Fixed

File: packages/core/core-flows/src/product/workflows/update-products.ts (hunk 2 above)
Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/update-products.ts:255]: Fixed

File: packages/core/core-flows/src/product/workflows/update-collections.ts (hunk 3 above)
Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/update-collections.ts]: Fixed

Comment (thetutlage) [reply packages/core/core-flows/src/product/workflows/update-collections.ts]: Fixed

**Review by sradevski — APPROVED** [review]
Really cool to support this, I think this should solve 80% of people's customization needs 👍

**Review by olivermrbl — APPROVED** [review]
Nice, excited about this! 🔥

**Comment by thetutlage** [comment]
There is real excitement for this 🔥

**Review by adrien2p — APPROVED** [review]

**Comment by olivermrbl** [comment]
@thetutlage, should we merge this? 🙏

**Comment by thetutlage** [comment]
Yeah. Have to update the PR description to list all hooks. I can do that post merging too

**Pull request merged**